src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+            PcntCountMode::Decrement => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE,
+        }
+    }
+}
+
+#[doc = " @brief PCNT channel action on control level"]
+#[derive(Debug, Copy, Clone)]
+pub enum PcntControlMode {
+    #[doc = "< Keep current count mode"]
+    Keep,
+    #[doc = "< Invert current count mode (increase -> decrease, decrease -> increase)"]
+    Reverse,
+    #[doc = "< Hold current count value"]
+    Disable,
+}
+
+impl Into<esp_idf_sys::pcnt_ctrl_mode_t> for PcntControlMode {
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        #[doc = "< PCNT watch point event: threshold1 value event"]
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        #[doc = "< PCNT watch point event: threshold0 value event"]
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        #[doc = "< PCNT watch point event: Minimum counter value"]
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        #[doc = "< PCNT watch point event: Maximum counter value"]
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        #[doc = "< PCNT watch point event: counter value zero event"]
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+pub struct PcntConfig<'d> {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<&'d AnyInputPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<&'d AnyInputPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
+}
+
+#[derive(Debug)]
+pub struct Pcnt {
+    unit: pcnt_unit_t,
+}
+
+impl<'d> Pcnt {
+    pub fn new() -> Result<Self, EspError> {
```

Comment:
Needs to take a `impl Peripheral<P = impl Pcnt>` + the same for each pin necessary, + the same for the channel, which seems also to be a peripheral.

Reply:
Can't pass the pins in new because they are channel specific.  I've renamed config() to channel_config() and the channel + pins + PcntChannelConfig are passed in there.